File: core-service/src/main/java/org/example/pantrypilot/model/Recipe.java
package org.example.pantrypilot.model;

import java.time.OffsetDateTime;
import java.util.ArrayList;
import java.util.List;

import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;
import jakarta.persistence.Table;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.ToString;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

@Entity
@Table(name = "recipe")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
@ToString(exclude = {"user", "ingredients"})
public class Recipe {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @Column(name = "title", nullable = false, length = 200)
    private String title;

    @Column(name = "instructions", nullable = false, columnDefinition = "text")
    private String instructions;

    @Column(name = "cook_time_minutes")
    private Integer cookTimeMinutes;

    @Column(name = "tags", columnDefinition = "text[]")
    private String[] tags;

    @Column(name = "source", length = 20)
    private String source;

    @CreationTimestamp
    @Column(name = "created_at", nullable = false, updatable = false)
    private OffsetDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at", nullable = false)
    private OffsetDateTime updatedAt;

    @Builder.Default
    @OneToMany(mappedBy = "recipe", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.LAZY)
    private List<RecipeIngredient> ingredients = new ArrayList<>();
}

File: frontend/src/types/recipe.ts
export interface RecipeIngredientResponse {
  id: number;
  name: string;
  quantity: number;
  unit: string | null;
  createdAt: string;
  updatedAt: string;
}

export interface RecipeResponse {
  id: number;
  title: string;
  instructions: string;
  cookTimeMinutes: number | null;
  tags: string[] | null;
  createdAt: string;
  updatedAt: string;
  ingredients: RecipeIngredientResponse[];
}

export interface RecipeSummaryResponse {
  id: number;
  title: string;
  cookTimeMinutes: number | null;
  tags: string[] | null;
  createdAt: string;
}

export interface CreateRecipeRequest {
  title: string;
  instructions: string;
  cookTimeMinutes?: number | null;
  tags?: string[] | null;
}

export interface UpdateRecipeRequest {
  title: string;
  instructions: string;
  cookTimeMinutes?: number | null;
  tags?: string[] | null;
}

export interface CreateRecipeIngredientRequest {
  name: string;
  quantity: number;
  unit?: string | null;
}

export interface UpdateRecipeIngredientRequest {
  name: string;
  quantity: number;
  unit?: string | null;
}
